src/librustc_borrowck/diagnostics.rs

```diff
@@ -314,6 +314,59 @@ let c = &i; // still ok!
 ```
 "##,
 
+E0504: r##"
+This error occurs when an attempt is made to move a borrowed variable into a
+closure. For example:
+
+```compile_fail
+use std::thread;
+
+struct FancyNum {
+    num: u8
+}
+
+fn main() {
+    let fancy_num = FancyNum { num: 5 };
+    let fancy_ref = &fancy_num;
+
+    thread::spawn(move || {
+        println!("child thread: {}", fancy_num.num);
+        // error: cannot move `fancy_num` into closure because it is borrowed
+    }).join().expect("thread failed to join");
+
+    println!("main thread: {}", fancy_ref.num);
+}
+```
+
+Here, `fancy_num` is borrowed by `fancy_ref` and so cannot be moved into
+the closure given to `thread::spawn`. There is no way to move a value into
+a closure while it is borrowed, as that would invalidate the borrow.
+
+If the value has to be borrowed, try limiting the lifetime of the borrow using
```

Comment:
Or they could use [`Arc`](https://doc.rust-lang.org/nightly/std/sync/struct.Arc.html) and [`RefCell`](https://doc.rust-lang.org/nightly/std/cell/struct.RefCell.html):

```Rust
use std::sync::Arc;
use std::cell::RefCell;

struct FancyNum {
    num: u8
}

fn main() {
    let fancy_num = FancyNum { num: 5 };
    let fancy_ref = Arc::new(RefCell::new(&fancy_num));
    let fancy_ref2 = fancy_ref.clone();

    let x = move || {
        println!("child thread: {}", fancy_ref2.borrow().num);
        // error: cannot move `fancy_num` into closure because it is borrowed
    };

    println!("main thread: {}", fancy_ref.borrow().num);
}
```